test/algorithms/simulation/sidb/detect_bdl_pairs.cpp

```diff
@@ -0,0 +1,301 @@
+//
+// Created by marcel on 21.07.23.
+//
+
+#include <catch2/catch_test_macros.hpp>
+
+#include <fiction/algorithms/simulation/sidb/detect_bdl_pairs.hpp>
+#include <fiction/types.hpp>
+
+using namespace fiction;
+
+TEST_CASE("Empty layout BDL detection", "[detect-bdl-pairs]")
+{
+    using layout = sidb_cell_clk_lyt_siqad;
+
+    const layout lyt{};
+
+    const auto result = detect_bdl_pairs(lyt, sidb_technology::cell_type::NORMAL);
+
+    CHECK(result.empty());
+}
+
+TEST_CASE("Atomic wire BDL detection", "[detect-bdl-pairs]")
+{
+    using layout = sidb_cell_clk_lyt_siqad;
+
+    layout lyt{{7, 0}, "Atomic wire"};
+
+    lyt.assign_cell_type({0, 0, 0}, sidb_technology::cell_type::INPUT);
+    lyt.assign_cell_type({1, 0, 0}, sidb_technology::cell_type::INPUT);
+
+    lyt.assign_cell_type({2, 0, 0}, sidb_technology::cell_type::NORMAL);
+    lyt.assign_cell_type({3, 0, 0}, sidb_technology::cell_type::NORMAL);
+    lyt.assign_cell_type({4, 0, 0}, sidb_technology::cell_type::NORMAL);
+    lyt.assign_cell_type({5, 0, 0}, sidb_technology::cell_type::NORMAL);
+
+    lyt.assign_cell_type({6, 0, 0}, sidb_technology::cell_type::OUTPUT);
+    lyt.assign_cell_type({7, 0, 0}, sidb_technology::cell_type::OUTPUT);
+
+    detect_bdl_pairs_params params{};
+
+    SECTION("default lower threshold")
+    {
+        const auto input_bdl_pairs  = detect_bdl_pairs(lyt, sidb_technology::cell_type::INPUT, params);
+        const auto output_bdl_pairs = detect_bdl_pairs(lyt, sidb_technology::cell_type::OUTPUT, params);
+        const auto normal_bdl_pairs = detect_bdl_pairs(lyt, sidb_technology::cell_type::NORMAL, params);
+
+        REQUIRE(input_bdl_pairs.empty());
+        REQUIRE(output_bdl_pairs.empty());
+        REQUIRE(normal_bdl_pairs.size() == 2);
+
+        const auto& normal_pair1 = normal_bdl_pairs[0];
+        const auto& normal_pair2 = normal_bdl_pairs[1];
+
+        CHECK(normal_pair1.type == layout::cell_type::NORMAL);
+        CHECK(normal_pair1.upper == cell<layout>{2, 0, 0});
```

Comment:
I think we cannot assume that normal_pair1 is the "first" normal pair. It could also be ``({3,0,0},{5,0,0})``.

Reply:
I guess if we are truly implementation independent, we cannot assume that indeed. It works because I know how `cell_level_layout` is implemented and that lower indices will be iterated over first. I might change the test.

Reply:
On my Apple M1 Pro, it actually takes the opposite order. So the test fails.

Reply:
That's really good to know! Thank you for trying this out. I will adjust the tests accordingly.